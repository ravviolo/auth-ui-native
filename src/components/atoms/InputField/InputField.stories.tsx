import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { StoryScreen } from 'storybook/StoryScreen';

import { InputField } from './InputField';

storiesOf('InputField', module)
  .addDecorator((getStory) => <StoryScreen>{getStory()}</StoryScreen>)
  .add('Text', () => (
    <InputField
      secureTextEntry={boolean('secureTextEntry', false)}
      testID={text('testID', 'input-field-test-id')}
    />
  ))
  .add('Password', () => (
    <InputField
      secureTextEntry={boolean('secureTextEntry', true)}
      testID={text('testID', 'input-field-test-id')}
    />
  ));
