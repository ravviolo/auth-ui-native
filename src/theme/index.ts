interface Colors {
  text: string;
  textLight: string;
  textWhite: string;
  primary: string;
  primaryDark: string;
  background: string;
}

export const colors: Colors = {
  text: '#3D3D3D',
  textLight: '#9E9E9E',
  textWhite: '#EDEDED',
  primary: '#FF4684',
  primaryDark: '#F9256C',
  background: '#fff',
};
