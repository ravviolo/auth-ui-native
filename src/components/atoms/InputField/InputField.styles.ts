import { StyleSheet } from 'react-native';
import { colors } from 'theme';

export const styles = StyleSheet.create({
  input: {
    borderColor: colors.textLight,
    borderRadius: 8,
    borderWidth: 1,
    height: 40,
    padding: 10,
  },
});
