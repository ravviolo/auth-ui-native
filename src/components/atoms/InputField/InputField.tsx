import { forwardRef, useImperativeHandle, useState } from 'react';
import { StyleSheet, TextInput } from 'react-native';
import { colors } from 'theme';

interface Props {
  testID: string;
  secureTextEntry?: boolean;
}

export interface TextInputHandle {
  getValue: () => string;
  clear: () => void;
}

export const InputField = forwardRef<TextInputHandle, Props>(
  ({ secureTextEntry = false, testID }, ref) => {
    const [value, setValue] = useState('');
    const [isFocused, setIsFocused] = useState(false);

    const toggleFocus = () => setIsFocused((focusState) => !focusState);

    useImperativeHandle(ref, () => ({
      getValue: () => value,
      clear: () => {
        setValue('');
      },
    }));

    return (
      <TextInput
        secureTextEntry={secureTextEntry}
        style={[
          styles.input,
          {
            borderColor: isFocused ? colors.text : colors.textLight,
          },
        ]}
        testID={testID}
        value={value}
        onBlur={toggleFocus}
        onChangeText={setValue}
        onFocus={toggleFocus}
      />
    );
  }
);

const styles = StyleSheet.create({
  input: {
    borderColor: colors.textLight,
    borderRadius: 8,
    borderWidth: 1,
    height: 40,
    padding: 10,
  },
});
