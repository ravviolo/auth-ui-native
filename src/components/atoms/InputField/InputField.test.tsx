import { act, fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';
import { colors } from 'theme';

import { InputField } from './InputField';

import type { Props } from './InputField';

describe('InputField', () => {
  const props: Props = {
    testID: 'input-field-test-id',
    secureTextEntry: false,
  };

  it('should render initially as empty input', () => {
    render(<InputField {...props} />);

    const inputField = screen.getByTestId('input-field-test-id');

    expect(inputField.props.value).toBe('');
  });

  it('should handle onChangeText events', () => {
    render(<InputField {...props} />);

    const inputField = screen.getByTestId('input-field-test-id');

    fireEvent.changeText(inputField, 'test input');

    expect(inputField.props.value).toBe('test input');
  });

  it('should change border color on input focus', async () => {
    render(<InputField {...props} />);

    const inputField = screen.getByTestId('input-field-test-id');

    expect(inputField).toHaveStyle({ borderColor: colors.textLight });

    await act(() => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      inputField.props.onFocus();
    });

    expect(inputField).toHaveStyle({ borderColor: colors.text });
  });

  it('should match snapshot', () => {
    const tree = testRenderer.create(<InputField {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
