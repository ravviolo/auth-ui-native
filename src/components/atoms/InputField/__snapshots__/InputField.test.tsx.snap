// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField should match snapshot 1`] = `
<TextInput
  onBlur={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  secureTextEntry={false}
  style={
    Array [
      Object {
        "borderColor": "#9E9E9E",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 40,
        "padding": 10,
      },
      Object {
        "borderColor": "#9E9E9E",
      },
    ]
  }
  testID="input-field-test-id"
  value=""
/>
`;
